File: app/src/main/java/com/lumiyaviewer/lumiya/slproto/messages/PickInfoUpdate.java
package com.lumiyaviewer.lumiya.slproto.messages;

import com.lumiyaviewer.lumiya.slproto.SLMessage;
import com.lumiyaviewer.lumiya.slproto.types.LLVector3d;
import java.nio.ByteBuffer;
import java.util.UUID;

/**
 * PickInfoUpdate
 * Update a pick.  ParcelID is set on the simulator as the message
 * passes through.
 * If TopPick is TRUE, the simulator will only pass on the message
 * if the agent_id is a god.
 * viewer -> simulator -> dataserver
 * reliable
 *
 * <p>Template: {@code PickInfoUpdate Low 185 NotTrusted Unencoded}
 * (recovered/reference/message_template.msg).
 */
public class PickInfoUpdate extends SLMessage {
    public AgentData AgentData_Field;
    public Data Data_Field;

    /** Block AgentData, Single. */
    public static class AgentData {
        public UUID AgentID; // LLUUID
        public UUID SessionID; // LLUUID
    }

    /** Block Data, Single. */
    public static class Data {
        public UUID CreatorID; // LLUUID
        public byte[] Desc; // Variable 2
        public boolean Enabled; // BOOL
        public byte[] Name; // Variable 1
        public UUID ParcelID; // LLUUID
        public UUID PickID; // LLUUID
        public LLVector3d PosGlobal; // LLVector3d
        public UUID SnapshotID; // LLUUID
        public int SortOrder; // S32
        public boolean TopPick; // BOOL
    }

    public PickInfoUpdate() {
        this.zeroCoded = false;
        this.AgentData_Field = new AgentData();
        this.Data_Field = new Data();
    }

    @Override
    public int CalcPayloadSize() {
        return this.Data_Field.Name.length + 50 + 2 + this.Data_Field.Desc.length + 16 + 24 + 4 + 1 + 36;
    }

    @Override
    public void Handle(SLMessageHandler sLMessageHandler) {
        sLMessageHandler.HandlePickInfoUpdate(this);
    }

    @Override
    public void PackPayload(ByteBuffer byteBuffer) {
        // Message number: Low 185 (PickInfoUpdate).
        byteBuffer.putShort((short) 0xFFFF);
        byteBuffer.put((byte) 0x00);
        byteBuffer.put((byte) 0xB9);
        packUUID(byteBuffer, this.AgentData_Field.AgentID);
        packUUID(byteBuffer, this.AgentData_Field.SessionID);
        packUUID(byteBuffer, this.Data_Field.PickID);
        packUUID(byteBuffer, this.Data_Field.CreatorID);
        packBoolean(byteBuffer, this.Data_Field.TopPick);
        packUUID(byteBuffer, this.Data_Field.ParcelID);
        packVariable(byteBuffer, this.Data_Field.Name, 1);
        packVariable(byteBuffer, this.Data_Field.Desc, 2);
        packUUID(byteBuffer, this.Data_Field.SnapshotID);
        packLLVector3d(byteBuffer, this.Data_Field.PosGlobal);
        packInt(byteBuffer, this.Data_Field.SortOrder);
        packBoolean(byteBuffer, this.Data_Field.Enabled);
    }

    @Override
    public void UnpackPayload(ByteBuffer byteBuffer) {
        this.AgentData_Field.AgentID = unpackUUID(byteBuffer);
        this.AgentData_Field.SessionID = unpackUUID(byteBuffer);
        this.Data_Field.PickID = unpackUUID(byteBuffer);
        this.Data_Field.CreatorID = unpackUUID(byteBuffer);
        this.Data_Field.TopPick = unpackBoolean(byteBuffer);
        this.Data_Field.ParcelID = unpackUUID(byteBuffer);
        this.Data_Field.Name = unpackVariable(byteBuffer, 1);
        this.Data_Field.Desc = unpackVariable(byteBuffer, 2);
        this.Data_Field.SnapshotID = unpackUUID(byteBuffer);
        this.Data_Field.PosGlobal = unpackLLVector3d(byteBuffer);
        this.Data_Field.SortOrder = unpackInt(byteBuffer);
        this.Data_Field.Enabled = unpackBoolean(byteBuffer);
    }
}

File: app/src/main/java/com/lumiyaviewer/lumiya/slproto/messages/PickInfoUpdate.kt
package com.lumiyaviewer.lumiya.slproto.messages

import com.lumiyaviewer.lumiya.slproto.SLMessage
import com.lumiyaviewer.lumiya.slproto.types.LLVector3d
import java.nio.ByteBuffer
import java.util.UUID

/**
 * PickInfoUpdate
 * Update a pick.  ParcelID is set on the simulator as the message
 * passes through.
 * If TopPick is TRUE, the simulator will only pass on the message
 * if the agent_id is a god.
 * viewer -> simulator -> dataserver
 * reliable
 *
 * <p>Template: {@code PickInfoUpdate Low 185 NotTrusted Unencoded}
 * (recovered/reference/message_template.msg).
 */
open class PickInfoUpdate : SLMessage() {
    @JvmField var AgentData_Field: AgentData = AgentData()
    @JvmField var Data_Field: Data = Data()

    /** Block AgentData, Single. */
    open class AgentData {
        @JvmField var AgentID: UUID? = null
        @JvmField var SessionID: UUID? = null
    }

    /** Block Data, Single. */
    open class Data {
        @JvmField var CreatorID: UUID? = null
        @JvmField var Desc: ByteArray? = null
        @JvmField var Enabled: Boolean = false
        @JvmField var Name: ByteArray? = null
        @JvmField var ParcelID: UUID? = null
        @JvmField var PickID: UUID? = null
        @JvmField var PosGlobal: LLVector3d? = null
        @JvmField var SnapshotID: UUID? = null
        @JvmField var SortOrder: Int = 0
        @JvmField var TopPick: Boolean = false
    }

    init {
        zeroCoded = false
    }

    override fun CalcPayloadSize(): Int {
        return Data_Field.Name!!.size + 50 + 2 + Data_Field.Desc!!.size + 16 + 24 + 4 + 1 + 36
    }

    override fun Handle(messageHandler: SLMessageHandler) {
        messageHandler.HandlePickInfoUpdate(this)
    }

    override fun PackPayload(byteBuffer: ByteBuffer) {
        // Message number: Low 185 (PickInfoUpdate).
        byteBuffer.putShort(0xFFFF.toShort())
        byteBuffer.put((0x00).toByte())
        byteBuffer.put((0xB9).toByte())
        packUUID(byteBuffer, AgentData_Field.AgentID)
        packUUID(byteBuffer, AgentData_Field.SessionID)
        packUUID(byteBuffer, Data_Field.PickID)
        packUUID(byteBuffer, Data_Field.CreatorID)
        packBoolean(byteBuffer, Data_Field.TopPick)
        packUUID(byteBuffer, Data_Field.ParcelID)
        packVariable(byteBuffer, Data_Field.Name, 1)
        packVariable(byteBuffer, Data_Field.Desc, 2)
        packUUID(byteBuffer, Data_Field.SnapshotID)
        packLLVector3d(byteBuffer, Data_Field.PosGlobal)
        packInt(byteBuffer, Data_Field.SortOrder)
        packBoolean(byteBuffer, Data_Field.Enabled)
    }

    override fun UnpackPayload(byteBuffer: ByteBuffer) {
        AgentData_Field.AgentID = unpackUUID(byteBuffer)
        AgentData_Field.SessionID = unpackUUID(byteBuffer)
        Data_Field.PickID = unpackUUID(byteBuffer)
        Data_Field.CreatorID = unpackUUID(byteBuffer)
        Data_Field.TopPick = unpackBoolean(byteBuffer)
        Data_Field.ParcelID = unpackUUID(byteBuffer)
        Data_Field.Name = unpackVariable(byteBuffer, 1)
        Data_Field.Desc = unpackVariable(byteBuffer, 2)
        Data_Field.SnapshotID = unpackUUID(byteBuffer)
        Data_Field.PosGlobal = unpackLLVector3d(byteBuffer)
        Data_Field.SortOrder = unpackInt(byteBuffer)
        Data_Field.Enabled = unpackBoolean(byteBuffer)
    }
}
